libpod/image/image.go

```diff
@@ -149,43 +210,344 @@ func (i *Image) ID() string {
 	}
-	return i.runtime.PullImage(name, options)
+	return "@" + digest.Hex(), nil
 }
 
-// Remove an image
-// This function is only complete enough for the stage 1 tests.
-func (i *Image) Remove(force bool) error {
-	_, err := i.runtime.RemoveImage(i.image, force)
-	return err
+// TagImage adds a tag to the given image
+func (i *Image) TagImage(tag string) error {
+	tags := i.Names()
+	if StringInSlice(tag, tags) {
+		return nil
+	}
+	tags = append(tags, tag)
+	return i.imageruntime.store.SetNames(i.ID(), tags)
+}
+
+// PushImage pushes the given image to a location described by the given path
+func (i *Image) PushImage(destination, manifestMIMEType, authFile, signaturePolicyPath string, writer io.Writer, forceCompress bool, signingOptions SigningOptions, dockerRegistryOptions *DockerRegistryOptions) error {
+	// PushImage pushes the src image to the destination
+	//func PushImage(source, destination string, options CopyOptions) error {
+	if destination == "" {
+		return errors.Wrapf(syscall.EINVAL, "destination image name must be specified")
+	}
+
+	// Get the destination Image Reference
+	dest, err := alltransports.ParseImageName(destination)
+	if err != nil {
+		if hasTransport(destination) {
+			return errors.Wrapf(err, "error getting destination imageReference for %q", destination)
+		}
+		// Try adding the images default transport
+		destination2 := DefaultTransport + destination
+		dest, err = alltransports.ParseImageName(destination2)
+		if err != nil {
+			// One last try with docker:// as the transport
+			destination2 = DefaultTransport + destination
+			dest, err = alltransports.ParseImageName(destination2)
+			if err != nil {
+				return errors.Wrapf(err, "error getting destination imageReference for %q", destination)
+			}
+		}
+	}
+
+	sc := GetSystemContext(signaturePolicyPath, authFile, forceCompress)
+
+	policyContext, err := getPolicyContext(sc)
+	if err != nil {
+		return err
+	}
+	defer policyContext.Destroy()
+
+	// Look up the source image, expecting it to be in local storage
+	src, err := is.Transport.ParseStoreReference(i.imageruntime.store, i.ID())
+	if err != nil {
+		return errors.Wrapf(err, "error getting source imageReference for %q", i.InputName)
+	}
+
+	copyOptions := GetCopyOptions(writer, signaturePolicyPath, nil, dockerRegistryOptions, signingOptions, authFile, manifestMIMEType, forceCompress)
+
+	// Copy the image to the remote destination
+	err = cp.Image(policyContext, dest, src, copyOptions)
+	if err != nil {
+		return errors.Wrapf(err, "Error copying image to the remote destination")
+	}
+	return nil
+}
+
+// MatchesID returns a bool based on if the input id
+// matches the image's id
+func (i *Image) MatchesID(id string) bool {
+	return strings.HasPrefix(i.ID(), id)
+}
+
+// ToStorageReference returns a *storageReference from an Image
+// TODO: cannot complete until we understand the return type which appears unexported
+func (i *Image) ToStorageReference() (error, error) {
```

Comment:
This can probably be private

Reply:
I want them exposed for external consumption.